File: 102LevelOrderTraversalofBinaryTree/LevelOrderTraversalofBinaryTree.py
# -*- coding: utf-8 -*-

import sys
import os
sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), "..")))
from node_structure.TreeNode import TreeNode
from typing import List

class LevelOrderTraversalofBinaryTree:
    def levelOrder(
        self,
        root: "TreeNode"
    ) -> "List[List[int]]":
        if not root: return []
        ans = []
        q = [root]
        level = 0
        while q:
            qsize = len(q)
            ans.append([])
            for _ in range(qsize):
                cur = q.pop(0)
                ans[level].append(cur.val)
                if cur.left: q.append(cur.left)
                if cur.right: q.append(cur.right)
            # print(ans)
            level += 1
        return ans


def main():
    test = LevelOrderTraversalofBinaryTree()
    print(test.levelOrder(TreeNode.fromList([1, 2, 3, 4, 5, 6, 7])))

if __name__ == "__main__":
    main()

File: 102LevelOrderTraversalofBinaryTree/LevelOrderTraversalofBinaryTree.go
package main

import (
	"fmt"
	nodestructure "leetcode/node_structure"
)

func levelOrder(root *nodestructure.TreeNode) [][]int {
	if nil == root {
		return [][]int{}
	}
	ans := [][]int{}
	q := []*nodestructure.TreeNode{}
	q = append(q, root)
	for len(q) > 0 {
		cur := []int{}
		qsize := len(q)
		for i := 0; i < qsize; i++ {
			tmp := q[0]
			q = q[1:]
			cur = append(cur, (*tmp).Val)
			if nil != tmp.Left {
				q = append(q, tmp.Left)
			}
			if nil != tmp.Right {
				q = append(q, tmp.Right)
			}
		}
		cp := append([]int(nil), cur...)
		ans = append(ans, cp)
	}
	return ans
}

func main() {
	fmt.Println("wait for TreeNode")
}
